tango/integrations/torch/eval.py

```diff
@@ -0,0 +1,209 @@
+
+        Parameters
+        ----------
+
+        model : :class:`Model`
+            The model to evaluate. It should return a ``dict`` from its ``forward()`` method
+            that includes all of the metrics in ``metric_names`` .
+        dataset_dict : :class:`~tango.common.dataset_dict.DatasetDictBase`
+            Should contain the test data.
+        dataloader : :class:`DataLoader`
+            The data loader that generates test batches. The batches should be :class:`dict`
+            objects.
+        test_split : :class:`str`, optional
+            The name of the data split used for evaluation in the ``dataset_dict``.
+            Default is "test".
+        seed : :class:`int`, optional
+            Used to set the RNG states at the beginning of training.
+        eval_steps : :class:`int`, optional
+            The number of steps to evaluate for. If not specified evaluation will
+            stop after a complete iteration through the ``dataloader``.
+        log_every : :class:`int`, optional
+            Log every this many steps. Default is ``1``.
+        device : ``Union[int, str, torch.device]``, optional
+            The device to evaluate on. Default to the first CUDA device if one is available,
+            otherwise CPU.
+        metric_names : ``Sequence[str]``, optional
+            The names of the metrics to track and aggregate. Default is ``("loss",)``.
+        auto_aggregate_metrics : :class:`bool`, optional
+            If ``True`` (the default), the metrics will be averaged across batches.
+            This may not be the correct behavior for some metrics (such as F1),
+            in which you should set this to ``False`` and handle the aggregation
+            internally in your model or with an :class:`EvalCallback`
+            (using :meth:`EvalCallback.post_batch()`).
+        callbacks : ``List[TrainCallback]``
+            A list of :class:`EvalCallback`.
+
+        """
+        # Set seeds.
+        random.seed(seed)
+        np.random.seed(seed)
+        torch.manual_seed(seed)
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(seed)
+
+        check_dataset(dataset_dict, test_split)
+
+        # Resolve device.
+        _device: torch.device
+        if device is None:
+            if torch.cuda.is_available():
+                print("CUDA is available")
+                _device = torch.device("cuda")
+            else:
+                _device = torch.device("cpu")
+        elif isinstance(device, int):
+            if device >= 0:
+                _device = torch.device(f"cuda:{device}")
+            else:
+                _device = torch.device("cpu")
+        elif isinstance(device, str):
+            _device = torch.device(device)
+        elif isinstance(device, torch.device):
+            _device = device
+        else:
+            raise ValueError(f"unexpected type for 'device': '{device}'")
+        device: torch.device = _device
```

Comment:
We should settle on a way to handle devices, and then put this into a utility function as well.

Reply:
7efcede